Better compute "sizePerShards" for the scroll-based-query

centic9 opened the pull request:
The "size" on a scroll query is not the actual max number of results for each paged-request, but rather the size *per Shard* for each paged-request. If you have a lot of shards (e.g. when you use wildcards in index-names to query data from many indices in combination with time-based indices-names), you likely end up requesting a huge number of results and will run into out of memory issues on the client.

This PR changes this logic by analyzing the number of shards that are affected by the query and limiting the overall scroll-size to be up to 5000 max instead of the 200-per-shard limit applied currently.

awislowski commented on src/main/java/pl/allegro/tech/search/elasticsearch/tools/reindex/process/QueryComponent.java:
```diff
@@ -36,12 +43,28 @@
+    GetSettingsResponse getSettingsResponse = client.admin().indices().getSettings(new GetSettingsRequest().indices(dataPointer.getIndexName())).actionGet();
+    int numShards = 0, numIndices = 0;
+    for(ObjectCursor<Settings> settings : getSettingsResponse.getIndexToSettings().values()) {
+      int replicas = settings.value.getAsInt("index.number_of_replicas", 0) + 1;
```
number of replicas does not play role in the size of results in the search request

awislowski commented on src/main/java/pl/allegro/tech/search/elasticsearch/tools/reindex/process/QueryComponent.java:
```diff
@@ -36,12 +43,28 @@
+    // find out how many indices and shards are affected by this query to not get huge result sets when there are very many indices affected by the name, e.g. when wildcards are used
+    // otherwise we regularly run into OOMs when a query goes against a large number of indices
+    // I did not find a better way to find out the number of Shards then to get a list of indices and for each index query the replicas/shards via the settings
+    GetSettingsResponse getSettingsResponse = client.admin().indices().getSettings(new GetSettingsRequest().indices(dataPointer.getIndexName())).actionGet();
```
Could you extract shardSize calculating into an other component and write a test for this method?

awislowski commented:
Thanks for the extension. I have added two comments to this code before merging.

centic9 commented:
I have updated the implementation, I refactored the computation as separate method in QueryComponent as it seems a bit much to have a separate class for this alone.
I also added a unit-test which verifies QueryComponent in general and this new calculation in particular.
